### PR TITLE
scheduler enhancements

### PR DESCRIPTION
**Pull request by santhoshb-msft**
1.  Added onetime meter event
2.  Fixed meter emit bug
3. Added meter on/off from app config
4. Added Runby column to scheduler

### CODE REVIEW

**Review by msalemcode — APPROVED** [review]

File: src/MeteredTriggerJob/MeteredTriggerHelper.cs [hunk 1]
```diff
@@ -22,68 +22,110 @@ public class Executor
...
+                        {
+                            TriggerSchedulerItem(scheduledItem,
+                                                 frequency,
+                                                 billingApiService,
+                                                 schedulerService,
+                                                 subscriptionUsageLogsRepository);
+                        }
+                        else
+                        {
+                            Console.WriteLine($"Item Id: {scheduledItem.Id} will not run as it doesnt match the time any difference. {_nextRunTime} UTC.");
```
Comment (tiffwchen) [inline src/MeteredTriggerJob/MeteredTriggerHelper.cs:109]: $"Item Id: {scheduledItem.Id} will not run as it doesn't match the time any difference. {_nextRunTime} UTC."

Added missing apostrophe

File: src/MeteredTriggerJob/MeteredTriggerHelper.cs [hunk 2]
```diff
@@ -22,68 +22,110 @@ public class Executor
...
+                        }
                     }
                 }
+                else
+                {
+                    Console.WriteLine($"{frequency} scheduled items will not be run as its disabled in the application config.");
```
Comment (tiffwchen) [inline src/MeteredTriggerJob/MeteredTriggerHelper.cs:115]: $"{frequency} scheduled items will not be run as it's disabled in the application config."

File: deployment/Templates/Deploy.ps1 [hunk 3]
```diff
@@ -193,12 +193,10 @@ Write-host "☁  Prepare publish files faS.Accelerator.or the web application"
...
-if ($MeteredSchedulerSupport -ne $true)
-{ 
-    Write-host "☁  Preparing the publish files for Metered Scheduler to Admin Site"
-    mkdir -p ..\..\Publish\AdminSite\app_data\jobs\triggered\MeteredTriggerJob
-    dotnet publish ..\..\src\MeteredTriggerJob\MeteredTriggerJob.csproj -c debug -o ..\..\Publish\AdminSite\app_data\jobs\triggered\MeteredTriggerJob  --runtime win-x64 --self-contained true 
-}
+
+Write-host "☁  Preparing the publish files for Metered Scheduler to Admin Site"
+mkdir -p ..\..\Publish\AdminSite\app_data\jobs\triggered\MeteredTriggerJob
+dotnet publish ..\..\src\MeteredTriggerJob\MeteredTriggerJob.csproj -c debug -o ..\..\Publish\AdminSite\app_data\jobs\triggered\MeteredTriggerJob  --runtime win-x64 --self-contained true
```
Comment (alexdrenea) [inline deployment/Templates/Deploy.ps1:199]: Do we know thy this needs to publish with --runtime and --self-contained? this won't work on Linux App Service plan (currently the rest of the solution works fine on Linux and it's actually beneficial as it's half the price of a windows ASP.

Comment (alexdrenea) [reply deployment/Templates/Deploy.ps1:199]: Oh, I just realized, webjobs are only supported in Windows App Service Plans.
I will open a new ticket to discuss alternatives.

**Comment by alexdrenea** [comment]
@santhoshb-msft some additional comments on the PR:
1. Can we add links to the Scheduled events page (/scheduler)? Something in the left side menu and maybe another grid item on the home page
2. Can we add an "edit" schedule page? Seems a bit counter-intuitive to ask users to just add a new scheduler with the same subscription/dimension in order to edit
3. Will be testing more, but ATM I can't seem to submit usage to the Metering API (getting BadRequest)

File: src/Services/Services/MeteredBillingAPIService.cs [hunk 4]
```diff
@@ -11,6 +11,7 @@ namespace Microsoft.Marketplace.SaaS.SDK.Services.Services
     using Microsoft.Marketplace.SaaS.SDK.Services.Configurations;
     using Microsoft.Marketplace.SaaS.SDK.Services.Contracts;
     using Microsoft.Marketplace.SaaS.SDK.Services.Models;
+    using Newtonsoft.Json;
```
Comment (alexdrenea) [inline src/Services/Services/MeteredBillingAPIService.cs:14]: Can we use System.Text.JSON instead? It's what we use in the rest of the project (or we should be aiming towards)

**Comment by santhoshbomma9** [comment]
@alexdrenea apologies didn't mention the scope of this PR, it is just to add onetime meter option + additional bug fixes. Any new features or enhancements, we will take them as new PR

**Comment by santhoshbomma9** [comment]
@code4clouds - given we have one approval, can I merge this in?

**Comment by code4clouds** [comment]
@santhoshb-msft - I have been swamped since last week and unable to find a focus time block to run my tests.  If I don't get to it by EOD Thursday, go ahead...

File: docs/Metered-Scheduler-Manager-Instruction.md [hunk 5]
```diff
@@ -5,6 +5,7 @@ Metered Scheduler Manager is a feature where Publisher can schedule **FIX Quanti
 1. Weekly
 1. Monthly
 1. Yearly
+1. OneTime
 ## Enable and Disable Metered Scheduler Manager Feature During Installation
 By defualt this feature is disabled and deployed as part of the installion. Publisher can enable this feature during the installion process by passing an optional parameter **MeteredSchedulerSupportEnabled** with **true** value as part of the installion script.
```
[the remark this answers is not in the document]
Comment (santhoshb-msft) [reply docs/Metered-Scheduler-Manager-Instruction.md]: fixed

File: src/AdminSite/Views/Scheduler/NewScheduler.cshtml [hunk 6]
```diff
@@ -65,6 +65,15 @@
         <form method="post" id="frmSchedulerEvents" asp-action="AddNewScheduledTrigger" asp-controller="Scheduler" >
             <div style="align-content:center">
                 <table>
+                    <tr>
+                        <th colspan="2">
+                            * Select atleast an hour later than current time as the current time is scheduler event is already passed<br />
```
[the remark this answers is not in the document]
Comment (santhoshb-msft) [reply src/AdminSite/Views/Scheduler/NewScheduler.cshtml]: fixed

File: src/AdminSite/Views/Scheduler/NewScheduler.cshtml [hunk 7]
```diff
@@ -65,6 +65,15 @@
...
+                            * To update an existing scheduler item please use the same Name, Subid, Plan and Dimension.
```
[the remark this answers is not in the document]
Comment (santhoshb-msft) [reply src/AdminSite/Views/Scheduler/NewScheduler.cshtml]: fixed

File: src/AdminSite/Views/Scheduler/SchedulerLogDetail.cshtml [hunk 8]
```diff
@@ -76,14 +76,18 @@
 
                 </table>
                 <br />
-
             </div>
+            <p>
+                This Log will show all the events on this subscription run by Manual or by any Scheduler item for this subscription.
```
[the remark this answers is not in the document]
Comment (santhoshb-msft) [reply src/AdminSite/Views/Scheduler/SchedulerLogDetail.cshtml]: fixed

File: src/MeteredTriggerJob/MeteredTriggerHelper.cs (hunk 1 above)
Comment (santhoshb-msft) [reply src/MeteredTriggerJob/MeteredTriggerHelper.cs]: fixed

File: src/MeteredTriggerJob/MeteredTriggerHelper.cs (hunk 2 above)
Comment (santhoshb-msft) [reply src/MeteredTriggerJob/MeteredTriggerHelper.cs]: fixed

File: src/Services/Services/MeteredBillingAPIService.cs (hunk 4 above)
Comment (santhoshb-msft) [reply src/Services/Services/MeteredBillingAPIService.cs:14]: 1,2  we can take as new enhancements
JSON -System.Text.JSON was having an issue, happy to take up as a PR for that

**Comment by santhoshb-msft** [comment]
@code4clouds - as discussed, this morning, merging this PR.

**Comment by santhoshb-msft** [comment]
Pasting the testing results reference
image

**Pull request merged**